images/virtualization-artifact/pkg/controller/vd/internal/source/object_ref_vi_on_pvc.go

```diff
@@ -0,0 +1,184 @@
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+     http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package source
+
+import (
+	"context"
+	"errors"
+	"fmt"
+
+	"k8s.io/apimachinery/pkg/api/resource"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	cdiv1 "kubevirt.io/containerized-data-importer-api/pkg/apis/core/v1beta1"
+
+	cc "github.com/deckhouse/virtualization-controller/pkg/controller/common"
+	"github.com/deckhouse/virtualization-controller/pkg/controller/service"
+	"github.com/deckhouse/virtualization-controller/pkg/controller/supplements"
+	"github.com/deckhouse/virtualization-controller/pkg/logger"
+	"github.com/deckhouse/virtualization-controller/pkg/util"
+	virtv2 "github.com/deckhouse/virtualization/api/core/v1alpha2"
+	"github.com/deckhouse/virtualization/api/core/v1alpha2/vicondition"
+)
+
+type ObjectRefVirtualImageOnPvc struct {
+	diskService *service.DiskService
+}
+
+func NewObjectRefVirtualImageOnPvc(diskService *service.DiskService) *ObjectRefVirtualImageOnPvc {
+	return &ObjectRefVirtualImageOnPvc{
+		diskService: diskService,
+	}
+}
+
+func (ds ObjectRefVirtualImageOnPvc) Sync(ctx context.Context, vd *virtv2.VirtualDisk, viRef *virtv2.VirtualImage, condition *metav1.Condition) (bool, error) {
+	log, _ := logger.GetDataSourceContext(ctx, objectRefDataSource)
+
+	supgen := supplements.NewGenerator(cc.VDShortName, vd.Name, vd.Namespace, vd.UID)
+	dv, err := ds.diskService.GetDataVolume(ctx, supgen)
+	if err != nil {
+		return false, err
+	}
+	pvc, err := ds.diskService.GetPersistentVolumeClaim(ctx, supgen)
+	if err != nil {
+		return false, err
+	}
+
+	switch {
+	case isDiskProvisioningFinished(*condition):
+		log.Info("Disk provisioning finished: clean up")
+
+		setPhaseConditionForFinishedDisk(pvc, condition, &vd.Status.Phase, supgen)
+
+		// Protect Ready Disk and underlying PVC.
+		err = ds.diskService.Protect(ctx, vd, nil, pvc)
+		if err != nil {
+			return false, err
+		}
+
+		err = ds.diskService.Unprotect(ctx, dv)
+		if err != nil {
+			return false, err
+		}
+
+		return CleanUpSupplements(ctx, vd, ds)
+	case cc.AnyTerminating(dv, pvc):
+		log.Info("Waiting for supplements to be terminated")
+	case dv == nil:
+		log.Info("Start import to PVC")
+
+		vd.Status.Progress = "0%"
+		vd.Status.SourceUID = util.GetPointer(viRef.GetUID())
+
+		refSupgen := supplements.NewGenerator(cc.VIShortName, viRef.Name, viRef.Namespace, viRef.UID)
+
+		refPvc, err := ds.diskService.GetPersistentVolumeClaim(ctx, refSupgen)
+		if err != nil {
+			return false, err
+		}
+
+		var size resource.Quantity
```

Comment:
The idea of using `var size resource.Quantity` here is to avoid shadowing the declaration of 'err'. However, the declaration of 'err' is already shadowed above at line 88.

Reply:
Sorry. I didn't understand what needs to be fixed

Reply:
Fixed.